src/Monitor.cs

```diff
@@ -381,18 +490,30 @@ private void device_OnPacketArrival(object sender, PacketCapture e)
             }
         }
 
+        private void DeserializeAndDispatchCompanion(byte[] buffer, Direction direction, uint sequenceNumber)
+        {
+            if (direction == Direction.Incoming)
+            {
+                _companionPacketDecoder.DecodeIncoming(buffer, sequenceNumber);
+            }
+            else if (direction == Direction.Outgoing)
+            {
+                _companionPacketDecoder.DecodeOutgoing(buffer, sequenceNumber);
+            }
+        }
+
         private void DeserializeAndDispatch(byte[] buffer, Direction direction)
```

Comment:
The reason I went with this is because the `DeserializeAndDispatch` became very convoluted to deal with two different streams of packets and I didn't find an elegant way to do that short of passing in another boolean flag or `source` type enum.
Probably it would be good to move the implementation of `DeserializeAndDispatch` into its own class too so that there is commonality of the dispatch strategy.

Reply:
Yes, I think extracting `DeserializeAndDispatch` into a separate class could make sense, but ultimately the events need to bubble back up to the event handlers in the `Monitor` class. It seems like that's one thing missing in `CompanionPacketDecoder` -- how does a consumer hook into those events?

Take a look at the demo project. I noticed it doesn't run because of missing DI dependencies. Can you fix that, and then illustrate how a consumer of the `Monitor` class would hook companion events?

I added the missing DI items to the demo project (didn't commit anything, tho) in order to test, and I'm getting a bunch of protobuf deserialization errors when I fire up Zwift and then start interacting with the companion app. Do you have an end-to-end test that works that I could see? Here's what my output looked like when I ran the demo program. The Protobuf errors were the result of me pressing buttons on the companion app, so there may be deserialization issues?

![image](https://user-images.githubusercontent.com/7749957/143113310-de425ec6-3a86-4bdb-a602-4584e8815ca9.png)

Reply:
There are still plenty things in reverse engineering that don’t work yet because not all messages are formed correctly so these errors appear unfortunately.

Additionally I think that I’ve found that the packet assembler sometimes discards protobuf messages because the assembled TCP packet contains more bytes than the length that is read. Trying to confirm that at the moment. 

I’ll have another look at the DI setup, I might have borked that in the last commits